From fcf87f3abc7467976f4729125f568de314b9ab0e Mon Sep 17 00:00:00 2001
From: Remi Duraffort <duraffortr@ocre.cea.fr>
Date: Wed, 30 Jun 2010 15:33:02 +0200
Subject: [PATCH] Fix ganestat (missing '}').

---
 src/tools/ganestat.pl |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/src/tools/ganestat.pl b/src/tools/ganestat.pl
index 00a17f6..79091fc 100755
--- a/src/tools/ganestat.pl
+++ b/src/tools/ganestat.pl
@@ -412,6 +412,7 @@ while (my $ligne=<STATS>)
         $fn_index ++;
       
       }
+    }
     elsif ( $tag =~ m/RQUOTA V. REQUEST/ )
     {
       next if ( ! ( $reste =~ m/^([^|]+)(.*)/ ) );  # go to next line
@@ -442,6 +443,7 @@ while (my $ligne=<STATS>)
         $fn_index ++;
       
       }
+   }
    elsif ( $tag =~ m/NLM V. REQUEST/ )
     {
       next if ( ! ( $reste =~ m/^([^|]+)(.*)/ ) );  # go to next line
-- 
1.7.1

