diff --git a/src/NFS_Protocols/nfs41_op_open.c b/src/NFS_Protocols/nfs41_op_open.c
index 47a3f2b..e34171c 100644
--- a/src/NFS_Protocols/nfs41_op_open.c
+++ b/src/NFS_Protocols/nfs41_op_open.c
@@ -112,7 +112,7 @@ int nfs41_op_open(struct nfs_argop4 *op, compound_data_t * data, struct nfs_reso
   int retval;
   fsal_name_t filename;
   bool_t AttrProvided = FALSE;
-  fsal_accessmode_t mode = 0000;
+  fsal_accessmode_t mode = 0600;
   nfs_fh4 newfh4;
   nfs_client_id_t nfs_clientid;
   nfs_worker_data_t *pworker = NULL;
diff --git a/src/NFS_Protocols/nfs4_op_create.c b/src/NFS_Protocols/nfs4_op_create.c
index cb251c4..57f568c 100644
--- a/src/NFS_Protocols/nfs4_op_create.c
+++ b/src/NFS_Protocols/nfs4_op_create.c
@@ -105,7 +105,7 @@ int nfs4_op_create(struct nfs_argop4 *op, compound_data_t * data, struct nfs_res
   cache_inode_status_t cache_status;
   int convrc = 0;
 
-  fsal_accessmode_t mode = 0000;
+  fsal_accessmode_t mode = 0600;
   fsal_name_t name;
 
   cache_inode_create_arg_t create_arg;
diff --git a/src/NFS_Protocols/nfs4_op_open.c b/src/NFS_Protocols/nfs4_op_open.c
index 426414d..500d076 100644
--- a/src/NFS_Protocols/nfs4_op_open.c
+++ b/src/NFS_Protocols/nfs4_op_open.c
@@ -112,7 +112,7 @@ int nfs4_op_open(struct nfs_argop4 *op, compound_data_t * data, struct nfs_resop
   int retval;
   fsal_name_t filename;
   bool_t AttrProvided = FALSE;
-  fsal_accessmode_t mode = 0000;
+  fsal_accessmode_t mode = 0600;
   nfs_fh4 newfh4;
   nfs_client_id_t nfs_clientid;
   nfs_worker_data_t *pworker = NULL;
