Signed-off-by: Frank Filz <ffilzlnx@us.ibm.com>
---
 src/MainNFSD/nfs_stats_snmp.c |   25 ++++++++++++++++---------
 1 files changed, 16 insertions(+), 9 deletions(-)

diff --git a/src/MainNFSD/nfs_stats_snmp.c b/src/MainNFSD/nfs_stats_snmp.c
index d4550ea..701a7f7 100644
--- a/src/MainNFSD/nfs_stats_snmp.c
+++ b/src/MainNFSD/nfs_stats_snmp.c
@@ -52,27 +52,34 @@ int get_snmpadm_conf(config_file_t in_config, external_tools_parameter_t * out_p
   int var_max, var_index;
   char *key_name;
   char *key_value;
+  config_item_t block;
 
-  /* this gets the bloc of your FS specific configuration */
-
-  blk_index = config_GetBlockIndexByName(in_config, CONF_SNMP_ADM_LABEL);
-
-  /* cannot read item */
-  if(blk_index < 0)
+  /* Get the config BLOCK */
+  if((block = config_FindItemByName(in_config, CONF_SNMP_ADM_LABEL)) == NULL)
+    {
+      DisplayLog("SNMP_ADM: Can't get label %s in file", CONF_SNMP_ADM_LABEL);
+      return ENOENT;
+    }
+  else if(config_ItemType(block) != CONFIG_ITEM_BLOCK)
     {
       DisplayLog("SNMP_ADM: Cannot read item \"%s\" from configuration file",
                  CONF_SNMP_ADM_LABEL);
+      /* Expected to be a block */
       return ENOENT;
     }
 
-  /* makes an iteration on the (key, value) couplets */
+  var_max = config_GetNbItems(block);
 
-  var_max = config_GetNbKeys(in_config, blk_index);
+  /* makes an iteration on the (key, value) couplets */
 
   for(var_index = 0; var_index < var_max; var_index++)
     {
+      config_item_t item;
+
+      item = config_GetItemByIndex(block, var_index);
+
       /* retrieve key's name */
-      err = config_GetKeyValue(in_config, blk_index, var_index, &key_name, &key_value);
+      err = config_GetKeyValue(item, &key_name, &key_value);
       if(err)
         {

