diff -upr nfs-ganesha-0.99.60.orig/ConfigParsing/conf_lex.c nfs-ganesha-0.99.60/ConfigParsing/conf_lex.c
--- nfs-ganesha-0.99.60.orig/ConfigParsing/conf_lex.c	2009-12-03 10:10:15.000000000 +0100
+++ nfs-ganesha-0.99.60/ConfigParsing/conf_lex.c	2009-12-03 17:29:12.000000000 +0100
@@ -574,7 +574,7 @@ char err_str[ERRLEN]="";
 char YY_PARSED_STRING[MAXSTRLEN];
 
 void YY_BUFFER_APPEND(char * s){
-    strncat( YY_PARSED_STRING,s, MAXSTRLEN);
+    strncat( YY_PARSED_STRING,s, MAXSTRLEN - strlen(YY_PARSED_STRING) - 1);
 }
 
 void YY_BUFFER_RESET(void){
diff -upr nfs-ganesha-0.99.60.orig/FSAL/FSAL_POSIX/DBExt/PGSQL/posixdb_add.c nfs-ganesha-0.99.60/FSAL/FSAL_POSIX/DBExt/PGSQL/posixdb_add.c
--- nfs-ganesha-0.99.60.orig/FSAL/FSAL_POSIX/DBExt/PGSQL/posixdb_add.c	2009-12-03 10:10:15.000000000 +0100
+++ nfs-ganesha-0.99.60/FSAL/FSAL_POSIX/DBExt/PGSQL/posixdb_add.c	2009-12-03 10:15:06.000000000 +0100
@@ -134,7 +134,7 @@ fsal_posixdb_status_t fsal_posixdb_add( 
    
     p_object_handle->ts = (int) time(NULL);
     p_object_handle->info = *p_object_info;
-    snprintf(handlets_str, MAX_HANDLEIDSTR_SIZE, "%i", p_object_handle->ts);
+    snprintf(handlets_str, MAX_HANDLETSSTR_SIZE, "%i", p_object_handle->ts);
     snprintf(nlink_str, MAX_NLINKSTR_SIZE, "%i", p_object_info->nlink);
     snprintf(ctime_str, MAX_CTIMESTR_SIZE, "%i", (int) p_object_info->ctime);
     snprintf(ftype_str, MAX_CTIMESTR_SIZE, "%i", (int) p_object_info->ftype);
diff -upr nfs-ganesha-0.99.60.orig/LRU/test_configurable_lru.c nfs-ganesha-0.99.60/LRU/test_configurable_lru.c
--- nfs-ganesha-0.99.60.orig/LRU/test_configurable_lru.c	2009-12-03 10:10:16.000000000 +0100
+++ nfs-ganesha-0.99.60/LRU/test_configurable_lru.c	2009-12-03 10:16:50.000000000 +0100
@@ -153,7 +153,10 @@ int do_new(  LRU_list_t * plru, int key 
   sprintf( tmpkey, "%d", key ) ;
   
   if( ( pentry = LRU_new_entry( plru, &status ) ) == NULL )
+  {
+    free( tmpkey ) ;
     return status ;
+  }
   
   pentry->buffdata.len   = strlen( tmpkey ) ;
   pentry->buffdata.pdata = tmpkey ;
diff -upr nfs-ganesha-0.99.60.orig/MainNFSD/nfs_stats_thread.c nfs-ganesha-0.99.60/MainNFSD/nfs_stats_thread.c
--- nfs-ganesha-0.99.60.orig/MainNFSD/nfs_stats_thread.c	2009-12-03 10:10:15.000000000 +0100
+++ nfs-ganesha-0.99.60/MainNFSD/nfs_stats_thread.c	2009-12-03 10:19:30.000000000 +0100
@@ -164,6 +164,7 @@ void * stats_thread( void * addr )
     {
       DisplayLog( "NFS STATS : Could not get inode for %s, no stats will be made...", 
 		  nfs_param.core_param.stats_file_path ) ;
+      fclose( stats_file ) ;
       return NULL ;
     }
   
diff -upr nfs-ganesha-0.99.60.orig/MainNFSD/Svc_auth_gss.c nfs-ganesha-0.99.60/MainNFSD/Svc_auth_gss.c
--- nfs-ganesha-0.99.60.orig/MainNFSD/Svc_auth_gss.c	2009-12-03 10:10:15.000000000 +0100
+++ nfs-ganesha-0.99.60/MainNFSD/Svc_auth_gss.c	2009-12-03 10:18:01.000000000 +0100
@@ -456,6 +456,7 @@ Gssrpc__svcauth_gss(struct svc_req *rqst
 		}
 		if ((gd = calloc(sizeof(*gd), 1)) == NULL) {
 			fprintf(stderr, "svcauth_gss: out_of_memory\n");
+			free(auth);
 			return (AUTH_FAILED);
 		}
 		auth->svc_ah_ops = &Svc_auth_gss_ops;
diff -upr nfs-ganesha-0.99.60.orig/shell/shell_utils.c nfs-ganesha-0.99.60/shell/shell_utils.c
--- nfs-ganesha-0.99.60.orig/shell/shell_utils.c	2009-12-03 10:10:16.000000000 +0100
+++ nfs-ganesha-0.99.60/shell/shell_utils.c	2009-12-03 10:20:53.000000000 +0100
@@ -321,9 +321,9 @@ int util_shell( int argc ,         /* IN
   
   for ( i = 1; i <argc ; i++ )
   {
-    strncat( command_line, argv[i], 1024 );
+    strncat( command_line, argv[i], strlen( command_line ) - 1024 - 1 );
     if ( i != argc - 1 )
-      strncat( command_line, " ", 1024 );
+      strncat( command_line, " ", strlen( command_line ) - 1024 - 1 );
   }
   
   /* launch the command */
diff -upr nfs-ganesha-0.99.60.orig/support/nfs_ip_stats.c nfs-ganesha-0.99.60/support/nfs_ip_stats.c
--- nfs-ganesha-0.99.60.orig/support/nfs_ip_stats.c	2009-12-03 10:10:16.000000000 +0100
+++ nfs-ganesha-0.99.60/support/nfs_ip_stats.c	2009-12-03 10:22:59.000000000 +0100
@@ -556,7 +556,10 @@ void nfs_ip_stats_dump( hash_table_t   *
              if( nfs_ip_stats_get( ht_ip_stats[j], 
                                    ipaddr, 
                                    &pnfs_ip_stats[j] ) != IP_STATS_SUCCESS )
-		return ;
+             {
+                fclose( flushipstat ) ;
+        	return ;
+             }
              ip_stats_aggreg.nb_call += (pnfs_ip_stats[j])->nb_call ;
 
 	     ip_stats_aggreg.nb_req_nfs2 += (pnfs_ip_stats[j])->nb_req_nfs2 ;
@@ -615,5 +618,7 @@ void nfs_ip_stats_dump( hash_table_t   *
 	 /* Check next client */
          RBT_INCREMENT(it);
        }
+
+       fclose( flushipstat ) ;
     }
 } /* nfs_ip_stats_dump */
