diff --git a/src/NFS_Protocols/nfs4_op_setattr.c b/src/NFS_Protocols/nfs4_op_setattr.c
index cfb2185..3dcb610 100644
--- a/src/NFS_Protocols/nfs4_op_setattr.c
+++ b/src/NFS_Protocols/nfs4_op_setattr.c
@@ -197,12 +197,16 @@ int nfs4_op_setattr(struct nfs_argop4 *op,
       /* Check for root access when using chmod */
       if(FSAL_TEST_MASK(sattr.asked_attributes, FSAL_ATTR_MODE))
         {
-          if(((sattr.mode & 0x800) &&
+          if(((sattr.mode & FSAL_MODE_SUID) &&
               ((data->pexport->options & EXPORT_OPTION_NOSUID) == EXPORT_OPTION_NOSUID))
-             || ((sattr.mode & 0x400)
+             || ((sattr.mode & FSAL_MODE_SGID)
                  && ((data->pexport->options & EXPORT_OPTION_NOSGID) ==
                      EXPORT_OPTION_NOSGID)))
             {
+              DisplayLogJdLevel(data->pclient->log_outputs, NIV_EVENT,
+                                "Setattr denied because setuid or setgid bit is disabled in configuration file."
+                                " setuid=%d, setgid=%d", sattr.mode & FSAL_MODE_SUID?1:0,
+                                sattr.mode & FSAL_MODE_SGID?1:0);
               res_SETATTR4.status = NFS4ERR_PERM;
               return res_SETATTR4.status;
             }
